### PR TITLE
feat: align OSS /info + on-demand fetch with cloud (Phase I)

### PR DESCRIPTION
**Pull request by even-wei**
**PR checklist**

- [x] Ensure you have added or ran the appropriate tests for your PR.
- [x] DCO signed

**What type of PR is this?**

Improvement (OSS side of DRC-3252 Phase I — full cloud/OSS API alignment)

**What this PR does / why we need it**:

OSS-side of the `/info` payload optimization that landed in the cloud PR ([recce-cloud-infra#1217](https://github.com/DataRecce/recce-cloud-infra/pull/1217)). Makes the OSS `/api/info` and `/api/model/{id}` contract match cloud, so both sides share the same frontend on-demand-fetch pattern.

**Backend:**

- `recce/server.py:574–600` — strip `raw_code`, `checksum`, and bulk `config` (keep `materialized`) from outbound `/api/info` lineage nodes at the API boundary. Done at the boundary rather than inside the adapter so internal consumers (`summary.py`, `get_change_analysis_cached`) still see the full cached lineage — zero internal regressions.
- `recce/adapter/dbt_adapter/__init__.py:488–498` — `DbtAdapter.get_model()` now includes `raw_code` so the existing `/api/model/{id}` endpoint can serve it on demand.

**Frontend (`js/packages/ui`):**

- `components/lineage/NodeSqlViewOss.tsx` — React Query fetch (`["modelDetail", nodeId]`, 5-min `staleTime`, 1 retry) when inline `raw_code` is `== null`. Skeleton during load. Inline `raw_code` wins when present (forward-compat for sources / cached payloads).
- `components/lineage/SandboxView.tsx` — new `isCodeLoading` prop; Skeleton during load; empty-state when no code; Run button disabled with tooltip when placeholder state would otherwise execute.
- `components/lineage/SandboxViewOss.tsx` — React Query fetch; editor state initializes only after `raw_code` resolves, never with placeholder text.
- `contexts/lineage/utils.ts` — removed the client-side checksum fallback block. `NodeChange` is now always computed server-side (cloud via DRC-3254, OSS via `DbtAdapter.get_change_analysis_cached`), so the fallback was unreachable.
- `NodeView.tsx` — code-change dot uses `changeStatus === "modified"` instead of `raw_code` comparison (verified equivalent in real-world states: in jaffle-shop-expand 1060 lineage-relevant nodes, 0 have weak checksums).

**Tests:**

- New `js/src/components/lineage/__tests__/NodeSqlView.ondemand.test.tsx` — 5 tests covering loading skeleton, successful fetch render, empty fallback, null-vs-undefined, inline-wins priority.
- Updated `NodeSqlView.test.tsx`, `lineage.test.ts`, `LineageView.test.tsx`, `contexts/lineage/__tests__/utils.test.ts` to pass a server-computed `diff` (since the client fallback was removed).

**Which issue(s) this PR fixes**:

- Resolves DRC-3252 (OSS side of Phase I)
- Closes DRC-3256 (original on-demand fetch scope)
- Closes DRC-3263 (absorbed after scope alignment request — no longer a separate follow-up)

**Special notes for your reviewer**:

- **Companion cloud PR:** [recce-cloud-infra#1217](https://github.com/DataRecce/recce-cloud-infra/pull/1217) — cloud-side strip + on-demand endpoint + `NodeDiff` precomputation.
- **Stripping at API boundary, not inside adapter:** `summary.py` and `get_change_analysis_cached` consume the same cached lineage dict internally and need `raw_code`/`checksum` present. Stripping at the API boundary (in `server.py`'s `/api/info` handler) preserves the internal contract. If you'd prefer the strip moved into the adapter, flag it — but the boundary approach is cheaper and lower-risk.
- **Columns kept inline (deliberate scope narrowing):** `NodeView` schema tab reads `columns` directly from lineage data. Stripping `columns` would require a parallel on-demand-fetch path through `NodeView` to avoid a regression. Follow-up noted; not blocking Phase I.
- **Tested locally:** ran OSS server against jaffle-shop-expand snowflake artifacts (1060 lineage-relevant nodes); on-demand fetch fires as expected when Code tab is opened.

**Verification:**
- Python: `make format && make flake8`: clean; `uv run pytest`: **1003 passed** (excluding `test_mcp_e2e.py` per repo convention).
- Frontend: `pnpm lint:fix && pnpm type:check`: clean; `pnpm exec vitest run`: **3686 passed, 5 skipped**; `pnpm run build`: succeeded.

**Does this PR introduce a user-facing change?**:

Yes — OSS users on large dbt projects see faster lineage-page loads due to smaller `/api/info` payload. Code tab uses on-demand fetch transparently (cache-on-first-open, 5-min `staleTime`).

### CODE REVIEW

**Comment by gcko** [comment]
## Code Review — PR #1320

### Summary

Solid, well-structured PR that strips `raw_code`/`checksum`/`config` from the `/api/info` lineage payload and adds on-demand fetching via `/api/model/{id}`. The architecture (strip at API boundary, not in adapter) is correct and the code is generally well-written. **No critical bugs found.** Three warnings worth discussing before merge.

### Findings

#### [Warning] `hasCodeChanges` semantic broadening may show false code-change dots

**File:** `js/packages/ui/src/components/lineage/NodeView.tsx:604`

**Issue:** The old `hasCodeChanges` compared `base.raw_code !== current.raw_code` directly — it showed the amber dot on the Code tab *only* when the SQL code actually differed. The new check `node.data.changeStatus === "modified"` relies on dbt's `state:modified` selector, which fires for config-only changes, description changes, macro changes, and other non-code modifications too.

This means a model that only has a config change (e.g., `materialized: table` → `materialized: view`) would now show the code-change dot even though `raw_code` is identical in both environments.

The PR description notes this was verified equivalent on jaffle-shop-expand (1060 nodes, 0 discrepancy), which is encouraging, but that's one project. In projects with config-heavy changes (e.g., toggling materialization strategies across environments), this could produce false positives on the Code tab indicator.

**Suggestion:** Consider renaming to `hasChanges` to reflect the broader semantics, or adding a comment noting the intentional broadening. Not blocking since the behavioral difference is minor (amber dot present vs absent), and the old behavior already missed config-only changes that arguably should show as "modified."

#### [Warning] No backend unit tests for `_strip_outbound_node` / `_strip_outbound_lineage`

**File:** `recce/server.py:574-596`

**Issue:** These two new functions sit on the critical `/api/info` response path — they transform every node in both `base` and `current` lineage dicts before sending to the client. There are no dedicated unit tests for them.

Edge cases that deserve test coverage:
- Node with no `config` key at all (sources, seeds without config)
- Node with `config: None` vs `config: {}` — both hit `node.get("config") or {}` but `None` and `{}` both evaluate falsy, so the behavior is correct — still, a test proves it
- Node missing `raw_code` and `checksum` (just verifying the dict comprehension doesn't break)
- Empty lineage dict (`{}`) and lineage with empty `nodes` (`{"nodes": {}}`)

The frontend tests cover the on-demand fetch flow well, and the existing integration tests pass, but the stripping logic itself is untested in isolation. If a future refactor touches this function, there's no regression safety net.

**Suggestion:** Add a small `test_strip_outbound_node` / `test_strip_outbound_lineage` test in `tests/test_server.py`. These are pure functions on dicts — easy to test.

#### [Warning] SandboxViewOss can overwrite user's deliberate empty editor state

**File:** `js/packages/ui/src/components/lineage/SandboxViewOss.tsx:110-120`

**Issue:** The render-time state sync at lines 110-120 seeds the editor with fetched `raw_code` when `modifiedCode === ""`. But the user might have deliberately cleared the editor buffer to type a completely new query. If the fetch resolves *after* the user clears the buffer, the condition `modifiedCode === ""` is true and the editor gets overwritten with the fetched code, undoing the user's intentional clear.

The window for this is narrow (fetch must resolve in the ~1-2 seconds between the user opening the sandbox and clearing the editor), but it's a real race condition in the user interaction flow.

**Suggestion:** Consider tracking a `userHasEdited` flag (set to `true` on any `setModifiedCode` call from user interaction, reset on dialog open) instead of relying on the empty-string heuristic. Alternatively, accept the risk given the narrow window — but add a code comment acknowledging the tradeoff.

### Verification Results

| Check | Result |
|-------|--------|
| Python tests (`make test`) | **1061 passed** |
| Python lint (`make flake8`) | **Clean** |
| Frontend tests (`pnpm test`) | **3686 passed, 5 skipped** |
| TypeScript (`pnpm type:check`) | **Clean** |
| Biome lint | **Clean** |

### What's Done Well

- **Stripping at API boundary** preserves the internal contract for `summary.py`, `get_change_analysis_cached`, and MCP consumers — correct architectural call.
- **Shallow copy, not mutation** of the cached lineage dicts avoids subtle cache-corruption bugs.
- **Loose equality (`== null`)** for the fetch trigger handles both `undefined` and JSON `null` — the right choice for Pydantic → JSON → TypeScript boundaries.
- **React Query dedup** means `NodeSqlViewOss` and `SandboxViewOss` share a single network request via the same `["modelDetail", nodeId]` cache key.
- **Comprehensive new test file** (`NodeSqlView.ondemand.test.tsx`) covers the major on-demand fetch scenarios.

### Verdict

Approved — no critical issues. The three warnings are non-blocking quality improvements that can be addressed in follow-ups. The implementation is well-reasoned, the test coverage is solid, and all CI checks pass.

**Review by gcko — APPROVED** [review]
Claude Code Review: No critical issues found. Three non-blocking warnings posted in review comment.

**Comment by gcko** [comment]
All three warnings addressed in f20c0ffc:

**1. `hasCodeChanges` semantic broadening** — Added a 5-line comment on `NodeView.tsx:604` explaining the intentional broadening from raw_code comparison to server-computed `changeStatus`, and why config-only changes showing the dot is acceptable.

**2. Backend unit tests for strip functions** — Added `TestStripOutboundNode` (7 tests) and `TestStripOutboundLineage` (4 tests) in `tests/test_server.py`. Covers: full node stripping, materialized preservation, missing/None/empty config, nodes without stripped fields, no-mutation guarantees, multi-node lineage, and empty lineage edge cases. This also addresses the Codecov gap (5 missing lines in `server.py`).

**3. SandboxViewOss editor overwrite race** — Replaced the `modifiedCode === ""` heuristic with a `userHasEdited` ref. The ref is set to `true` on any user edit via `handleUserEdit` callback, reset to `false` when the dialog opens. The render-time sync now checks `!userHasEdited.current` instead of the empty-string heuristic, so fetched code never overwrites deliberate user changes.

All checks pass: 29 Python tests (11 new), 3686 frontend tests, TypeScript clean, Biome clean, flake8 clean.

File: js/packages/ui/src/components/lineage/NodeSqlView.tsx [hunk 1]
```diff
@@ -125,6 +125,10 @@ export const NodeSqlView = ({
   const modelName =
     node.data.data.base?.name ?? node.data.data.current?.name ?? "";
 
+  // Defensive: show "No code available" only when raw_code is missing from
+  // all sources (forward-compatibility for when backend strips raw_code).
+  const hasCode = original !== undefined || modified !== undefined;
```
[the remark this answers is not in the document]
Comment (gcko) [reply js/packages/ui/src/components/lineage/NodeSqlView.tsx]: Fixed in e375997c. Switched to loose equality (`!= null`) so both `null` and `undefined` raw_code are treated as missing, correctly rendering the "No code available" state.

File: recce/server.py [hunk 2]
```diff
@@ -593,6 +618,14 @@ async def get_info():
 
     state_metadata = context.state_loader.state.metadata if context.state_loader.state else None
     lineage_diff = context.get_lineage_diff()
+    # DRC-3263: strip large/redundant fields from outbound lineage nodes.
+    # The underlying cached dicts are not mutated; _strip_outbound_lineage
+    # returns a shallow copy with stripped node dicts.
+    outbound_lineage = type(lineage_diff)(
+        base=_strip_outbound_lineage(lineage_diff.base),
+        current=_strip_outbound_lineage(lineage_diff.current),
+        diff=lineage_diff.diff,
```
[the remark this answers is not in the document]
Comment (gcko) [reply recce/server.py]: Fixed in e375997c. Switched to `lineage_diff.model_copy(update={...})` which avoids Pydantic v2 re-validation while preserving the immutability of the cached lineage dicts.

File: js/packages/ui/src/components/lineage/SandboxViewOss.tsx [hunk 3]
```diff
@@ -40,30 +41,91 @@ interface SandboxViewProps {
...
+  } else if (
+    isOpen &&
+    resolvedRawCode !== prevResolvedRawCode &&
+    !userHasEdited.current
+  ) {
+    // Fetch resolved after the dialog was already open — seed the editor
+    // with the freshly resolved raw_code, but only if the user hasn't
+    // edited the modified buffer.
+    setPrevResolvedRawCode(resolvedRawCode);
+    setModifiedCode(resolvedRawCode ?? "");
```
[the remark this answers is not in the document]
Comment (gcko) [reply js/packages/ui/src/components/lineage/SandboxViewOss.tsx:128]: No change needed — this is React's recommended "store previous props in state" pattern (the documented replacement for `getDerivedStateFromProps`). The React docs explicitly recommend this over `useEffect` to avoid an extra re-render cycle. The pattern at lines 112-129 compares current vs previous prop values and synchronizes derived state during render, which React supports and optimizes for. The `userHasEdited` ref (added in f20c0ffc) already prevents the race condition mentioned.
